src/ralph/backends/http/lrs.py

```diff
@@ -0,0 +1,106 @@
+
+import logging
+from typing import List, Union
+from urllib.parse import urlparse
+
+import httpx
+from httpx import Client, HTTPError
+from more_itertools import chunked
+from pydantic import HttpUrl, parse_obj_as
+
+from ralph.conf import settings
+from ralph.exceptions import BackendException
+
+from .base import BaseHTTP, HTTPBackendStatus
+
+lrs_settings = settings.BACKENDS.HTTP.LRS
+logger = logging.getLogger(__name__)
+
+
+class LRSHTTP(BaseHTTP):
+    """LRS HTTP backend."""
+
+    name = "lrs"
+
+    def __init__(
+        self,
+        connection_url: str = lrs_settings.CONNECTION_URL,
+        headers: dict = lrs_settings.HEADERS,
+    ):
+        """Instantiates the LRS client.
+
+        Args:
+            connection_url (HttpUrl): Connection URL to a LRS.
+            headers (json): Headers defined for the LRS server connection
+        """
+        self.connection_url = parse_obj_as(HttpUrl, connection_url)
+        self.auth = (self.connection_url.user, self.connection_url.password)
+        self.headers = headers
+        self.origin = (
+            urlparse(self.connection_url).scheme
+            + "://"
+            + urlparse(self.connection_url).netloc
+        )
+
+    def status(self):
+        """Implements HTTP backend check for server status."""
+        try:
+            response = httpx.get(self.origin + "/__heartbeat__")
+            response.raise_for_status()
+        except HTTPError:
+            return HTTPBackendStatus.ERROR
+
+        return HTTPBackendStatus.OK
+
+    def get(self, target: str, chunk_size: Union[None, int] = None):
+        """Get statements from LRS.
+
+        The `get` method defined in the LRS spefication returns `statements` array and
+        `more` IRL. The latter is required when the `statement` list returned has been
+        limited.
+        """
+
+        def fetch_statements(url, origin=self.origin):
+            with Client(auth=self.auth, headers=self.headers) as client:
+                response = client.get(url=origin + url)
+                response.raise_for_status()
+                statements = response.json()["statements"]
+                for statement in statements:
+                    yield statement
+                if "more" in response.json() and response.json()["more"]:
+                    yield from fetch_statements(url=response.json()["more"])
+
+        try:
+            yield from fetch_statements(target)
+        except HTTPError as error:
+            msg = "Failed to fetch statements."
+            logger.error("%s. %s", msg, error)
+            raise BackendException(msg, *error.args) from error
+
+    def post(
+        self,
+        target: str,
+        data: List[Union[bytes, dict]],
+        chunk_size: Union[None, int] = 500,
+        ignore_errors: bool = False,
+    ) -> int:
+        """Post statements to LRS.
+
+        Stores in an LRS a statement or a set of statements.
+        """
+        statements = 0
```

Comment:
I agree with the naming change, but I wonder if it is the appropriate output for this function, as I am not sure how this would be used.

As is, the function will either raise an Exception, or return the number of statements provided by the user (which the user already has access to). Returning a count would seem more appropriate if we would also return the count even in the event of an exception (something like `{"status": "error", "statements_posted_count": 1000; "remaining_statements_count": 150}`), but this would prevent us from raising an error.

Maybe we could return the output of the requests, in case the LRS provides feedback ?

Reply:
The objective is to be able to use Ralph as a FastAPI server that can GET and POST statements from other LRS. If you look at the POST statements route [here](https://github.com/openfun/ralph/blob/639e2456a44b9acd0e196a57dbf4567f9b063fc7/src/ralph/api/routers/statements.py#L349), we use the number of statements posted for logging information.